chalk-integration/src/lowering.rs

```diff
@@ -132,15 +137,25 @@ impl<'k> Env<'k> {
                     actual: 0,
                 });
             } else {
-                return Ok(chalk_ir::TyData::Function(chalk_ir::Fn {
-                    num_binders: k.binders.len(interner),
+                return Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
+                    name: chalk_ir::TypeName::FnDef(id.clone()),
                     substitution: chalk_ir::Substitution::empty(interner),
                 })
                 .intern(interner)
                 .cast(interner));
             }
         }
 
+        if let Some(id) = self.closure_ids.get(&name.str) {
+            return Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
```

Comment:
I think we're missing the check for `k.binders.len() != 0` here?

Reply:
Also this code looks like it should maybe be factored and shared better -- I presume this match arm covers only the case where there were no type parameters supplied. That probably didn't matter when lowering was simpler ...

Reply:
I must be missing the case where type parameters are supplied...

Reply:
Yeah, I wasn't handling this correctly.